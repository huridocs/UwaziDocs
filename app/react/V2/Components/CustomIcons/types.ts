interface CustomIconProps {
  className?: string;
}

export type { CustomIconProps };
