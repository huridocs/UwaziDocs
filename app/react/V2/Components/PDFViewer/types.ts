import { TextSelection } from 'react-text-selection-handler/dist/TextSelection';

type TextHighlight = {
  key: string;
  textSelection: TextSelection;
  color?: string;
};

export type { TextHighlight };
