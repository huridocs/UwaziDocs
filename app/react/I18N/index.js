import I18NLink from './components/I18NLink';
import I18N from './components/I18N';
import I18NUtils from './utils';
import I18NApi from './I18NApi';
import * as actions from './actions/I18NActions';

export { I18NLink, I18NUtils, I18N, I18NApi, actions };
export { I18NMenu } from './components/I18NMenu';
