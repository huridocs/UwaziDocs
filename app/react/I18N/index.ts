import I18NLink from './components/I18NLink';
import I18NUtils from './utils';
import I18NApi from './I18NApi';

export { I18NApi };
export { I18N } from './components/I18N';
export { I18NMenu } from './components/I18NMenu';
export * as actions from './actions/I18NActions';
export { I18NLink as I18NLinkV2 } from './I18NLinkV2';
export { Translate } from './Translate';
export { t } from './translateFunction';
export { TranslateModal } from './TranslateModal';
export { I18NLink, I18NUtils };
