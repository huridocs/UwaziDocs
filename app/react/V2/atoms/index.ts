export { atomStore } from './store';
export { notificationAtom } from './notificationAtom';
export { settingsAtom } from './settingsAtom';
export { templatesAtom } from './templatesAtom';
export { translationsAtom } from './translationsAtom';
export { thesaurisAtom } from './thesaurisAtom';
export { globalMatomoAtom } from './globalMatomoAtom';
export { userAtom } from './userAtom';
export type { AtomStoreData } from './store';
export type { notificationAtomType } from './notificationAtom';
