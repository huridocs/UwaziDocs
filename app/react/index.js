import React from 'react';
import ReactDOM from 'react-dom';

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

import { AppContainer } from 'react-hot-loader';

import { SENTRY_APP_DSN } from 'app/config';
import App from './App';

import './App/sockets';

if (SENTRY_APP_DSN) {
  Sentry.init({
    dsn: SENTRY_APP_DSN,
    integrations: [new BrowserTracing()],
    tracesSampleRate: 0.1,
  });
}

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
