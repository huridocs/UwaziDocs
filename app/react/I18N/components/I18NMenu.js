import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'UI';
import { actions, Translate, t } from 'app/I18N';

import { DropdownList } from 'app/Forms';

const prepareDropdownValues = (languageMap, locale, user) => {
  const languages = languageMap.toJS();

  const selectedLanguage =
    languages.find(lang => lang.key === locale) || languages.find(lang => lang.default);

  if (user.get('_id') && user.get('role') !== 'collaborator') {
    languages.push({ label: 'Live translate', key: 'livetranslate', type: 'livetranslate' });
  }

  return { languages, selectedLanguage };
};
const listItem = (item, i18nmode) => {
  if (!item.type) {
    return <span>{item.label}</span>;
  }

  return (
    <div className="live-translate">
      <Icon icon="circle" className={i18nmode ? 'live-on' : 'live-off'} />
      <Translate>{item.label}</Translate>
    </div>
  );
};
class I18NMenu extends Component {
  static reload(url) {
    window.location.href = url;
  }

  render() {
    const {
      languages: languageMap,
      locale,
      location,
      i18nmode,
      toggleInlineEdit,
      user,
    } = this.props;
    const { languages, selectedLanguage } = prepareDropdownValues(languageMap, locale, user);

    if (location.search.match(/page=/)) {
      const cleanSearch = location.search.split(/page=\d+|&page=\d+/).join('');
      location.search = cleanSearch === '?' ? '' : cleanSearch;
    }

    const regexp = new RegExp(`^/?${locale}/|^/?${locale}$`);
    const path = location.pathname.replace(regexp, '/');

    if (languageMap.size === 0) {
      return null;
    }

    return languageMap.count() > 1 || user.size ? (
      <div className="menuNav-I18NMenu" role="navigation" aria-label="Languages">
        {!i18nmode && (
          <DropdownList
            data={languages}
            defaultValue={selectedLanguage}
            textField="label"
            className="menuNav-language"
            itemComponent={({ item }) => listItem(item)}
            valueComponent={({ item }) => listItem(item)}
            onSelect={selected => {
              if (selected.type === 'livetranslate') {
                toggleInlineEdit();
              } else {
                const url = `/${selected.key}${path}${
                  path.match('document') ? '' : location.search
                }`;
                I18NMenu.reload(url);
              }
            }}
          />
        )}
        {i18nmode && (
          <div className="menuNav-language">
            <button
              className="singleItem"
              type="button"
              onClick={toggleInlineEdit}
              aria-label={t('System', 'Turn off inline translation', null, false)}
            >
              <div className="live-translate">
                <Icon icon="circle" className={i18nmode ? 'live-on' : 'live-off'} />
              </div>
            </button>
            <span className="singleItem">
              <Translate>Live translate</Translate>
            </span>
          </div>
        )}
      </div>
    ) : (
      <div className="menuNav-language">
        <span className="singleItem">{selectedLanguage.label}</span>
      </div>
    );
  }
}

I18NMenu.defaultProps = {
  locale: null,
};

I18NMenu.propTypes = {
  location: PropTypes.instanceOf(Object).isRequired,
  languages: PropTypes.instanceOf(Object).isRequired,
  user: PropTypes.instanceOf(Object).isRequired,
  i18nmode: PropTypes.bool.isRequired,
  toggleInlineEdit: PropTypes.func.isRequired,
  locale: PropTypes.string,
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ toggleInlineEdit: actions.toggleInlineEdit }, dispatch);
}

export function mapStateToProps(state) {
  return {
    languages: state.settings.collection.get('languages'),
    i18nmode: state.inlineEdit.get('inlineEdit'),
    locale: state.locale,
    user: state.user,
  };
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(I18NMenu));
