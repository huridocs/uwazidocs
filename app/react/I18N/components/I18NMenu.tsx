import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import React, { Component } from 'react';

import { NeedAuthorization } from 'app/Auth';
import { actions, t } from 'app/I18N';

const defaultProps = {
  locale: null,
  handleClick: () => {},
};

export type I18NMenuProps = typeof defaultProps & {
  languages: any;
  location: any;
  i18nmode: any;
  toggleInlineEdit: any;
};

export type I18NMenuState = {
  i18NMenuVisibility: boolean;
};

class I18NMenu extends Component<I18NMenuProps, I18NMenuState> {
  static reload(url: string) {
    window.location.href = url;
  }

  static defaultProps = defaultProps;

  constructor(props: I18NMenuProps) {
    super(props);

    this.state = { i18NMenuVisibility: false };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    this.setState({ i18NMenuVisibility: true });
  }

  render() {
    const { languages, locale, location, i18nmode, toggleInlineEdit } = this.props;

    let path = location.pathname;
    const regexp = new RegExp(`^/?${locale}/|^/?${locale}$`);
    path = path.replace(regexp, '/');

    const selectedLanguageLabel = languages.map((lang: any) => {
      const key = lang.get('key');
      if (key === locale) {
        const label = lang.get('label');
        return label;
      }
      return '';
    });

    return (
      <div className="translation-menu">
        <NeedAuthorization roles={['admin', 'editor']}>
          <label className="translation-switch">
            <input
              type="checkbox"
              className="translation-switch-input"
              aria-label={t('System', 'Add/edit translations', null, false)}
              onChange={toggleInlineEdit}
              checked={!!i18nmode}
            />
            <span className="translation-switch-slider" />
          </label>
          {/* <button
            className={`menuNav-btn btn btn-default${i18nmode ? ' inlineEdit active' : ''}`}
            type="button"
            onClick={toggleInlineEdit}
            aria-label={t('System', 'Add/edit translations', null, false)}
          >
            <Icon icon="language" size="lg" />
          </button> */}
        </NeedAuthorization>
        <p>
          <svg
            width="16"
            height="14"
            viewBox="0 0 16 14"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              // eslint-disable-next-line max-len
              d="M6.86134 7.38519L8.59317 9.09656L8.075 10.4841L5.95453 8.36359L2.54543 11.7727L1.58064 10.8079L5.04769 7.38177C4.19544 6.43405 3.5136 5.38403 3.01588 4.27267H4.3795C4.7954 5.08405 5.32041 5.8511 5.9545 6.55678C6.93632 5.46585 7.65565 4.21812 8.11587 2.90222H0.499969V1.54541H5.27269V0.181763H6.63634V1.54541H11.4091V2.91245H9.41135C8.9136 4.52155 8.06816 6.04199 6.88179 7.36473L6.86134 7.38519ZM11.0682 5.63634H12.4318L15.5 13.8182H14.1363L13.3693 11.7727H10.1307L9.36362 13.8182H7.99997L11.0682 5.63634ZM11.75 7.45337L10.642 10.4091H12.858L11.75 7.45337Z"
              fill="white"
            />
          </svg>
          {selectedLanguageLabel}
        </p>
        <ul className="menuNav-I18NMenu" role="navigation" aria-label="Languages">
          {languages.count() > 1 &&
            languages.map((lang: any) => {
              const key = lang.get('key');
              const label = lang.get('label');
              const url = `/${key}${path}${path.match('document') ? '' : location.search}`;
              return (
                <li className={`menuNav-item${locale === key ? ' is-active' : ''}`} key={key}>
                  <a className="menuNav-btn btn btn-default" href={url} aria-label={label}>
                    {label}
                  </a>
                </li>
              );
            })}
        </ul>
      </div>
    );
  }
}

export function mapStateToProps(state: any) {
  return {
    languages: state.settings.collection.get('languages'),
    i18nmode: state.inlineEdit.get('inlineEdit'),
    locale: state.locale,
  };
}

function mapDispatchToProps(dispatch: any) {
  return bindActionCreators({ toggleInlineEdit: actions.toggleInlineEdit }, dispatch);
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(I18NMenu));
