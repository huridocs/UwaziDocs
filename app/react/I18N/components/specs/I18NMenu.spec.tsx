/**
 * @jest-environment jsdom
 */
import React from 'react';
import { act, fireEvent, RenderResult, screen, render } from '@testing-library/react';
import { Location, MemoryRouter } from 'react-router-dom';
import { createStore, Provider } from 'jotai';
import { ClientUserSchema } from 'app/apiResponseTypes';
import { inlineEditAtom, localeAtom, settingsAtom, userAtom } from 'V2/atoms';
import { TestAtomStoreProvider } from 'V2/testing';
import { UserRole } from 'shared/types/userSchema';
import { LanguageISO6391 } from 'shared/types/commonTypes';
import { I18NMenu } from '../I18NMenu';

const defaultLanguages = [
  {
    _id: '1',
    label: 'English',
    key: 'en' as LanguageISO6391,
    localized_label: 'English',
    default: true,
  },
  {
    _id: '2',
    label: 'Spanish',
    key: 'es' as LanguageISO6391,
    localized_label: 'Espa単ol',
    default: false,
  },
];

const users = [
  { _id: 'admin', username: 'admin', role: UserRole.ADMIN, email: '' },
  { _id: 'collab', username: 'collab', role: UserRole.COLLABORATOR, email: '' },
];

describe('I18NMenu', () => {
  const initialEntry: Partial<Location> = { pathname: '/library' };
  const inlineEditAtomValue = { inlineEdit: false };
  let renderResult: RenderResult;
  let settingsAtomValue = { languages: defaultLanguages };
  let localeAtomValue = 'en';

  Reflect.deleteProperty(global.window, 'location');
  window.location = { ...window.location, assign: jest.fn() };

  const renderComponent = (user?: ClientUserSchema) => {
    renderResult = render(
      <MemoryRouter initialEntries={[initialEntry]}>
        <TestAtomStoreProvider
          initialValues={[
            [settingsAtom, settingsAtomValue],
            [userAtom, user],
            [localeAtom, localeAtomValue],
            [inlineEditAtom, inlineEditAtomValue],
          ]}
        >
          <I18NMenu />
        </TestAtomStoreProvider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    initialEntry.pathname = '/library';
    localeAtomValue = 'en';
    settingsAtomValue = { languages: defaultLanguages };
    inlineEditAtomValue.inlineEdit = false;
    jest.clearAllMocks();
  });

  it('should render the links to the different languages', () => {
    renderComponent();
    const links = screen.getAllByRole('link');
    expect(links.map(link => link.getAttribute('href'))).toEqual(
      expect.arrayContaining(['/en/library', '/es/library'])
    );
  });

  it('should not render anything if there is only one language', () => {
    settingsAtomValue = {
      languages: [
        { _id: '2', label: 'Spanish', key: 'es', localized_label: 'Espa単ol', default: true },
      ],
    };
    renderComponent();
    const links = screen.queryAllByRole('link');
    expect(links.length).toBe(0);
  });

  it('should show as active the current locale', async () => {
    renderComponent(users[0]);
    const [listItem] = renderResult
      .getAllByRole('listitem')
      .filter(item => item.textContent === 'English');
    expect(listItem.getAttribute('class')).toBe('menuNav-item active');
  });

  it('should active toggle translation edit mode when clicking Live translate', async () => {
    renderComponent(users[0]);
    expect(renderResult.container).toMatchSnapshot('before turning on live translate');
    await act(async () => {
      fireEvent.click(screen.getByText('Live translate').parentElement!);
    });
    expect(renderResult.container).toMatchSnapshot('after turning on live translate');
  });

  describe('when there is a user', () => {
    it('should render then laguages and the live translate option', () => {
      renderComponent(users[0]);
      expect(renderResult.getByText('Live translate')).toBeInTheDocument();
    });

    it('should not render live translate for unauthorized users', () => {
      renderComponent(users[1]);
      expect(renderResult.queryByText('Live translate')).not.toBeInTheDocument();
    });

    it('should display the language section if there is only one language', () => {
      settingsAtomValue = {
        languages: [
          { _id: '2', label: 'Spanish', key: 'es', localized_label: 'Espa単ol', default: true },
        ],
      };
      renderComponent(users[1]);
      const links = screen.queryAllByRole('link');
      expect(links.length).toBe(1);
      expect(links.map(link => link.getAttribute('href'))).toEqual(
        expect.arrayContaining(['/es/library'])
      );
    });
  });

  describe('Paths', () => {
    it.each`
      locale  | currentPath                | search                        | expectedPath
      ${'es'} | ${'/es/documents'}         | ${'?search'}                  | ${'/documents'}
      ${'en'} | ${'/en/entity/r2dzptt7ts'} | ${'?page=2'}                  | ${'/entity/r2dzptt7ts'}
      ${null} | ${'/templates/2452345'}    | ${'?query=weneedmoreclerics'} | ${'/templates/2452345?query=weneedmoreclerics'}
      ${null} | ${'/entity/2452345'}       | ${'?query=weneedmoreclerics'} | ${'/entity/2452345?query=weneedmoreclerics'}
      ${'es'} | ${'/es/templates/2452345'} | ${'?query=weneedmoreclerics'} | ${'/templates/2452345?query=weneedmoreclerics'}
    `(
      'should create the expected links for $pathName',
      async ({ locale, currentPath, search, expectedPath }) => {
        localeAtomValue = locale;
        initialEntry.pathname = currentPath;
        initialEntry.search = search;
        renderComponent(users[0]);
        const links = screen.getAllByRole('link');
        expect(links.map(link => link.getAttribute('href'))).toEqual(
          expect.arrayContaining([`/en${expectedPath}`, `/es${expectedPath}`])
        );
      }
    );
  });

  describe('reloading after language change', () => {
    const testStore = createStore();
    testStore.set(userAtom, users[0]);
    testStore.set(localeAtom, 'en');
    testStore.set(settingsAtom, settingsAtomValue);

    it('should trigger a reload if the current language is deleted', async () => {
      const result = render(
        <MemoryRouter initialEntries={[initialEntry]}>
          <Provider store={testStore}>
            <I18NMenu />
          </Provider>
        </MemoryRouter>
      );

      const newSettingsAtomValue = {
        languages: [
          {
            _id: '2',
            label: 'Spanish',
            key: 'es' as LanguageISO6391,
            localized_label: 'Espa単ol',
            default: true,
          },
        ],
      };

      await act(() => {
        testStore.set(settingsAtom, newSettingsAtomValue);
      });

      result.rerender(
        <MemoryRouter initialEntries={[initialEntry]}>
          <Provider store={testStore}>
            <I18NMenu />
          </Provider>
        </MemoryRouter>
      );

      expect(window.location.assign).toHaveBeenCalledTimes(1);
      expect(window.location.assign).toHaveBeenCalledWith('/library');
    });
  });
});
