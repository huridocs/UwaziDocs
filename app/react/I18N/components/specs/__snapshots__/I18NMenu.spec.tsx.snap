// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18NMenu should active toggle translation edit mode when clicking Live translate: after turning on live translate 1`] = `
<div>
  <li
    aria-label="Languages"
    class="menuNav-I18NMenu"
  >
    <div
      class="menuNav-language"
    >
      <button
        aria-label="Turn off inline translation"
        class="singleItem"
        type="button"
      >
        <div
          class="live-translate"
        >
          <svg
            height="20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="10"
              cy="10"
              fill="#88eacd"
              r="7"
            />
          </svg>
        </div>
      </button>
      <span
        class="singleItem"
      >
        <span
          class="translation active "
        >
          Live translate
        </span>
      </span>
    </div>
  </li>
</div>
`;

exports[`I18NMenu should active toggle translation edit mode when clicking Live translate: before turning on live translate 1`] = `
<div>
  <li
    aria-label="Languages"
    class="menuNav-I18NMenu"
  >
    <div
      class="menuNav-language"
    >
      <button
        class="singleItem dropdown"
        type="button"
      >
        <span>
          English
        </span>
         
        <svg
          aria-hidden="true"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.22 8.22a.75.75 0 0 1 1.06 0L10 11.94l3.72-3.72a.75.75 0 1 1 1.06 1.06l-4.25 4.25a.75.75 0 0 1-1.06 0L5.22 9.28a.75.75 0 0 1 0-1.06Z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <ul
        class="dropdown-menu"
      >
        <li
          class="menuNav-item active"
        >
          <a
            href="/en/library"
          >
            English
          </a>
        </li>
        <li
          class="menuNav-item"
        >
          <a
            href="/es/library"
          >
            Español
          </a>
        </li>
        <li
          class="menuNav-item"
        >
          <button
            class="live-translate"
            type="button"
          >
            <svg
              height="20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="10"
                cy="10"
                fill="#ffe66b"
                r="7"
              />
            </svg>
            <span
              class="translation "
            >
              Live translate
            </span>
          </button>
        </li>
      </ul>
    </div>
  </li>
</div>
`;
